src/python/pants/bsp/spec.py

```diff
@@ -0,0 +1,359 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Any
+
+# -----------------------------------------------------------------------------------------------
+# Base types
+# -----------------------------------------------------------------------------------------------
+
+Uri = str
+
+
+@dataclass(frozen=True)
+class BuildTargetIdentifier:
+    """A unique identifier for a target, can use any URI-compatible encoding as long as it is unique
+    within the workspace.
+
+    Clients should not infer metadata out of the URI structure such as the path or query parameters,
+    use BuildTarget instead.
+    """
+
+    # The target’s Uri
+    uri: Uri
+
+    @classmethod
+    def from_json_dict(cls, d):
+        return cls(uri=d["uri"])
+
+
+@dataclass(frozen=True)
+class BuildTargetCapabilities:
+    # This target can be compiled by the BSP server.
+    can_compile: bool
+
+    # This target can be tested by the BSP server.
+    can_test: bool
+
+    # This target can be run by the BSP server.
+    can_run: bool
+
+    # This target can be debugged by the BSP server.
+    can_debug: bool
+
+    @classmethod
+    def from_json_dict(cls, d):
+        return cls(
+            can_compile=d["canCompile"],
+            can_test=d["canTest"],
+            can_run=d["canRun"],
+            can_debug=d["canDebug"],
+        )
+
+
+@dataclass(frozen=True)
+class BuildTarget:
+    """Build target contains metadata about an artifact (for example library, test, or binary
+    artifact)"""
```

Comment:
To make clear, "Target" here is a BSP concept, not Pants targets, right? Would be good to clarify in docstring.

Reply:
The BSP concept is "build target." The term "target" isn't used individually. Will add comment in a follow-up.

Reply:
I do like that this file is called `bsp.spec`, makes that more clear. I was going to propose calling this `BspBuildTarget` - I still think it might be worth it, but fine either way.